### PR TITLE
COVERAGE: Allow component management

### PR DESCRIPTION
**Pull request by neolynx**
Fixes #

## Requirements

All new code should be covered with tests, documentation should be updated. CI should pass.

## Checklist

- [ ] unit-test added (if change is algorithm)
- [ ] functional test added/updated (if change is functional)
- [ ] man page updated (if applicable)
- [ ] bash completion updated (if applicable)
- [ ] documentation updated
- [ ] author name in `AUTHORS`

### CODE REVIEW

File: cmd/publish_source_drop.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	return err
+}
+
+func makeCmdPublishSourceDrop() *commander.Command {
+	cmd := &commander.Command{
+		Run:       aptlyPublishSourceDrop,
+		UsageLine: "drop <distribution>",
+		Short:     "drops revision of published repository",
+		Long: `
+Command drops revision of a published repository.
```
Comment (neolynx) [inline cmd/publish_source_drop.go:50]: let's not mention revisions externally, let's use the term 'stages changes" ?

File: cmd/publish_source_add.go [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+		context.Progress().Printf("Adding component %q with source %q [%s]...\n", component, name, published.SourceKind)
+
+		sources[component] = names[i]
+	}
+
+	err = collectionFactory.PublishedRepoCollection().Update(published)
+	if err != nil {
+		return fmt.Errorf("unable to save to DB: %s", err)
+	}
+
```
Comment (neolynx) [inline cmd/publish_source_add.go:58]: add message: you can now run aptly publish...

similar to the message shown on mirror create, snapshot create, ...

File: cmd/publish_source_drop.go (hunk 1 above)
Comment (cfiehe) [reply cmd/publish_source_drop.go]: Ok, it was a relic. Fixed.

File: api/publish.go [hunk 3]
```diff
@@ -403,3 +492,367 @@ func apiPublishDrop(c *gin.Context) {
...
+
+	b.Component = component
+	b.Name = revision.Sources[component]
+
+	if c.Bind(&b) != nil {
+		return
+	}
+
+	if b.Component != component {
+		delete(sources, component)
```
Comment (neolynx) [inline api/publish.go:713]: need test coverage ?

File: deb/publish.go [hunk 4]
```diff
@@ -73,6 +83,192 @@ type PublishedRepo struct {
...
+			Component: component,
+			Name:      name,
+		})
+	}
+
+	return sourceList
+}
+
+func (revision *PublishedRepoRevision) SourceNames() []string {
+	sources := revision.Sources
```
Comment (neolynx) [inline deb/publish.go:152]: need test coverage

File: deb/publish.go [hunk 5]
```diff
@@ -281,14 +477,28 @@ func NewPublishedRepo(storage, prefix, distribution string, architectures []stri
 	return result, nil
 }
 
-// MarshalJSON requires object to filled by "LoadShallow" or "LoadComplete"
-func (p *PublishedRepo) MarshalJSON() ([]byte, error) {
-	type sourceInfo struct {
-		Component, Name string
+func (revision *PublishedRepoRevision) MarshalJSON() ([]byte, error) {
```
Comment (neolynx) [inline deb/publish.go:480]: need test coverage

Comment (cfiehe) [reply deb/publish.go:480]: Test has been added.

File: deb/publish.go (hunk 4 above)
Comment (cfiehe) [reply deb/publish.go:152]: Test has been added.

File: api/publish.go (hunk 3 above)
Comment (cfiehe) [reply api/publish.go:714]: Test has been added.

**Pull request closed without merging**